Add documentation about Alternative Text for Images (#225)
primer/design: content/accessibility/alternative-text-for-images.mdx

---

> title: Alternative Text for Images

Reviewer: ```suggestion
title: Alternative text for images
```

I only know this because it was called out in my PRs 😅 [Sentence case is preferred for titles](https://primer.style/contribute/documentation#grammar-and-usage)

---

## Overview

> Text alternatives are a primary way for making information accessible because they can be rendered through any sensory modality (for example, visual, auditory or tactile) to match the needs of the user. Providing text alternatives allows the information to be rendered in a variety of ways by a variety of user agents. For example, a person who cannot see a picture can have the text alternative read aloud using synthesized speech.
>

> > Source: https://www.w3.org/TR/UNDERSTANDING-WCAG20/text-equiv-all.html

Reviewer: Do we need the `>` here? Can we just have this as regular text and not quoted text, or was there a reason for doing it this way?

Author: Well, this is a literal quote from https://www.w3.org/TR/UNDERSTANDING-WCAG20/text-equiv-all.html, that's why the quote block is used. Logically and semantically should be correct. Let me know if you disagree or you want me to rephrase the paragraph so it's not literal quoting anymore

## Why?

> Images can make the web user experience better and more pleasant, especially for users with cognitive and learning disabilities. But not all images used in websites are accessible, and that can be a major barrier. Accessible images are beneficial for everyone, especially for people using screen reader and other assistive technologies, as well as is beneficial for Search Engine Optimization (SEO). 

Reviewer: ```suggestion
Images can make the web user experience better and more pleasant, especially for users with cognitive and learning disabilities. Not all images used in websites are accessible, and that can be a major barrier. Accessible images are beneficial for everyone, especially for people using screen readers and other assistive technologies, as well Search Engine Optimization (SEO). 
```

You can read more about why alternative text for images is important in the [Web Accessibility Tutorials Guidance on how to create websites that meet WCAG
](https://www.w3.org/WAI/tutorials/images/#why-is-this-important).

## Guidelines & Examples

> ### Informative & Decorative images

Reviewer: ```suggestion
### Informative and decorative images
```

> To help you determine the best way to communicate the information of an image to a screen reader users you can use [the alt text decision tree provided by the W3C](

Reviewer: ```suggestion
To help you determine the best way to communicate the information of an image to a screen reader user, you can use [the alt text decision tree provided by the W3C](
```

https://www.w3.org/WAI/tutorials/images/decision-tree/).

> Depending the purpose of the images, they can be categorized into two categories: [informative](https://www.w3.org/WAI/tutorials/images/informative/) and [decorative](https://www.w3.org/WAI/tutorials/images/decorative/) images.

Reviewer: ```suggestion
Depending the purpose of the images, they can be categorized into one of two categories: [informative](https://www.w3.org/WAI/tutorials/images/informative/) or [decorative](https://www.w3.org/WAI/tutorials/images/decorative/) images.
```

> - Informative images convey a simple concept or information that can be expressed in a short phrase or sentence. The alt text must contain the description of what is displayed visually in the image.

Reviewer: ```suggestion
- Informative images convey a simple concept or information that can be expressed in a short phrase or sentence. The `alt` text must contain the description of what is displayed visually in the image.
```

- Decorative images don’t add information to the content of a page. For example an image that is included to make the website more visually attractive. You will always get an affirmative answer to the question "if the image was removed, would the user still get all the information from the page?".

### For designers

When you create a new design and it includes images, include captions of a description for each of them. Annote which images are informative and which ones are decorative.

### For engineers

> As you're developing a new component, feature, or page that contains an image, use proper [semantic html](https://primer.style/design/accessibility/semantic-html) and use the `<img>` element with a proper alternative text using the `alt` attribute.

Reviewer: ```suggestion
As you're developing a new component, feature, or page that contains an image, use proper [semantic HTML](https://primer.style/design/accessibility/semantic-html) and use the `<img>` element with proper alternative text using the `alt` attribute.
```

> Every `<img>` element must have an `alt` attribute. It could have content for informative images or be empty for decorative images, but it must always be present.

Reviewer: Could we add the code for an empty `alt` attribute just to be explicit?

### How to write good alt text

> A missing alt attribute is bad, but an alt with wrong or useless information is not a better solution. When writing the alt text of an image you should always have in mind some rules:

Reviewer: ```suggestion
A missing `alt` attribute is bad, but an `alt` attribute with wrong or useless information is not a better solution. When writing the `alt` text of an image, you should always have in mind some rules:
```

> - Keep it short. It's usually better to avoid long, boring, or irrelevant information. The alt text should be the most concise description possible of the image’s purpose.

Reviewer: ```suggestion
- Keep it short. It's usually better to avoid long, boring, or irrelevant information. The `alt` text should be the most concise description possible of the image’s purpose.
```

> - The alternative text should not include the words button, link, or image. Screen readers already provide that information.

Reviewer: Good call on this one!

- In [textual images](https://www.w3.org/WAI/tutorials/images/textual/

> ), images that contains words that are important to understanding the content, the words are included in the text alternative.

Reviewer: ```suggestion
) (images that contains words that are important to understanding the content), the words are included in the text alternative.
```

> - Don't fill the alt text with keywords. The alt attribute is often misused as a place to store SEO keywords or copyright information.

Reviewer: ```suggestion
- Don't fill the `alt` text with keywords. The `alt` attribute is often misused as a place to store SEO keywords or copyright information.
```

> - Combine alt text with the caption or the content. If the image is sufficiently described in the text — for example, a simple diagram illustrating what's written in the web page text — it can have brief alt text such as "Diagram of work flow as describe above".

Reviewer: ```suggestion
- Combine `alt` text with the caption or the content. If the image is sufficiently described in the text, for example, a simple diagram illustrating what's written in the web page text, it can have brief `alt` text such as "Diagram of work flow described above".
```

You can read more tips in the [Web Accessibility Tutorials Guidance on how to create websites that meet WCAG
](https://www.w3.org/WAI/tutorials/images/tips/).

## Additional resources

- [WebAIM Alternative Text](https://webaim.org/techniques/alttext/)

> - [Deep Dive: Making your images accessible](https://github.com/github/accessibility/blob/main/docs/deep-dive-notes/2022-01-26-images.md)

Reviewer 2: The `Deep dive` won't be accessible to those outside of GitHub org. Should we remove this url, or add a note?

Author: Yes, I thought the same, but I saw @lindseywild did link it in here document as well. But I agree with you, this documentation is public, so not everyone will be able to access that links.

@inkblotty @lindseywild any idea? Shall we delete the internal links or add a note?

Reviewer: Ah, whoops. I would say we keep it, but have a note that it's internal only! It's super helpful stuff 💪 I can update my doc as well!

Author: Thanks, I will take the advantage and include the note in your document as well :)

Reviewer: You rock! Thanks so much!

](https://www.w3.org/WAI/tutorials/images/#why-is-this-important).

## Guidelines and examples

### Informative and decorative images

To help you determine the best way to communicate the information of an image to a screen reader user, you can use [the alt text decision tree provided by the W3C](
https://www.w3.org/WAI/tutorials/images/decision-tree/).

Depending the purpose of the images, they can be categorized into one of two categories: [informative](https://www.w3.org/WAI/tutorials/images/informative/) or [decorative](https://www.w3.org/WAI/tutorials/images/decorative/) images.

- Informative images convey a simple concept or information that can be expressed in a short phrase or sentence. The `alt` text must contain the description of what is displayed visually in the image.
- Decorative images don’t add information to the content of a page. For example, an image that is included to make the website more visually attractive. You will always get an affirmative answer to the question "if the image was removed, would the user still get all the information from the page?".

### For designers

> When you create a new design and it includes images, include captions of a description for each of them. Annote which images are informative and which ones are decorative.

Reviewer 3: ```suggestion
When you create a new design and it includes images, include captions with a description for each of them. Annotate which images are informative and which ones are decorative.
```

Is this what you meant to say?

### For engineers

As you're developing a new component, feature, or page that contains an image, use proper [semantic HTML](https://primer.style/design/accessibility/semantic-html) and use the `<img>` element with proper alternative text using the `alt` attribute.

Every `<img>` element must have an `alt` attribute. It could have content for informative images or be empty for decorative images, but it must always be present.

### How to write good alt text

A missing `alt` attribute is bad, but an `alt` attribute with wrong or useless information is not a better solution. When writing the `alt` text of an image, you should always have in mind some rules:

> - Keep it short. It's usually better to avoid long, boring, or irrelevant information. The `alt` text should be the most concise description possible of the image’s purpose.

Reviewer 3: ```suggestion
- Keep it short. Avoid long, boring, or irrelevant information. The `alt` text should be the most concise description possible of the image’s purpose.
```

- The alternative text should not include the words button, link, or image. Screen readers already provide that information.
- In [textual images](https://www.w3.org/WAI/tutorials/images/textual/

> ) (images that contains words that are important to understanding the content), the words are included in the text alternative.

Author: Yes :)

- Don't fill the `alt` text with keywords. The `alt` attribute is often misused as a place to store SEO keywords or copyright information.

> - Combine `alt` text with the caption or the content. If the image is sufficiently described in the text, for example, a simple diagram illustrating what's written in the web page text, it can have brief `alt` text such as "Diagram of work flow described above".

Reviewer 3: ```suggestion
- Combine `alt` text with the caption or the content. If the image is sufficiently described in the text (for example, a simple diagram illustrating what's written in the web page text) it can have a brief `alt` text such as "Diagram of work flow described above".
```

Is it OK to say "above"?
